### ISSUE TITLE
Make \ProvideHook independent of package loading order

### ISSUE THREAD

**Issue by jlaurens** [latex3/latex2e#601]
## Brief outline of the enhancement

Actually `\NewHook` following `\ProvideHook` raises an error.

## Minimal example showing the current behaviour

```latex
\documentclass{article}
% Fake loading of package A:
\NewHook{A}
\ProvideHook{B}
% Fake loading of package B:
\NewHook{B}     % <---- ERROR
\ProvideHook{A}
\begin{document}
\end{document}
```
Whatever package loading order we choose, there is always an error.

## Discussion

We can perfectly replace here each `\NewHook` by `\ProvideHook` and the problem disappears.
But then what is the real advantage of `\NewHook` versus `\ProvideHook`?

There is a real advantage with the `\newcommand`/`\providecommand` design because it allows a default meaning of the command. But with hooks, there is nothing equivalent to the meaning. Moreover, there is no way to say if hooks were declared with `\NewHook` or `\ProvideHook`.

BTW, `\IfHookExistsTF` is of greater interest.

It is asked by the documentation to favor `\NewHook` versus `\ProvideHook` but the benefits are not clear.
Actually, what would be the difference in CTAN packages if `\NewHook` was implemented like `\ProvideHook`?

## Solution

One very simple solution would be to make `\NewHook` behave exactly like `\ProvideHook`, of course including reverse ordering when required.

A new name for that unique command could be `\DeclareHook` (and the same for L3) which would be consistent with the actual documentation ('declare' is used very often for hooks, including in section titles) and the internal `\__hook_if_declared:nTF`.

I largely prefer `\ActivateHook` which refers to the hook being active or not. This would allow to really clarify the documentation. An example for `\AddToHook`,

"The hook doesn’t have to exist for code to be added to it."

(which is a very weird concept of existence) would become 

"The hook doesn’t have to be active for code to be added to it."

Another one for `\UseHook`.

"Execute the hook code inside a command or environment."

would become

"If the hook is active, execute all the hook code chunks otherwise does nothing"

and `\IfHookExistsTF` would become `\IfHookActiveTF`.

Replace Active with Declare if you wish but it is far less clear.

**Comment by FrankMittelbach** [comment]
Your issue makes it clear that chosing `\ProvideHook` as a name was a very bad choice as it implies a close connection to the concept of `\newcommand` and `\providecommand` and it isn't. 

Hook are global data structures and they should no be declared in different places (for different purposes). It therefore essential that if two independent packages declare the hook "foo"  the resulting clash should be detected up front which is why they should all should `\NewHook`. And there is no reason why one package should "declare" or "provide the declaration" for another package's hook, on the contrary (which is what is wrong with your discussion example above, it is not an acceptable/supported use case).

So making `\NewHook` work like current `\ProvideHook`, i.e allowing it to be used several times for the same hook name is not the answer, nor that people start to use `\ProvideHook` when they should use `\NewHook`.

Bottom line `\ProvideHook` needs to vanish again. What will come instead we'll see, maybe `\ActivateGenericHook`
because that is the real use case why `\ProvideHook` was added.

**Comment by FrankMittelbach** [comment]
`\ProvideHook` is deprecated in upcoming release --- new approach documented and implemented.
